feat: Add configuration to disable global TCP/UDP intercepts

bgruszka opened the pull request:
## Description

In shared development environments, multiple developers need to work on the same service simultaneously. Currently, when one developer creates a global TCP/UDP intercept, it blocks all traffic to that port, preventing other developers from creating their own intercepts on the same service. This creates bottlenecks in team workflows.

This PR add a configuration option to restrict intercepts to only HTTP-based intercepts with header or path filters, allowing multiple developers to create personal intercepts simultaneously.

  ## Changes

  ### New Helm Chart Configuration

  Added `intercept.allowGlobalIntercepts` configuration option:

  ```yaml
  intercept:
    allowGlobalIntercepts: true  # default
```

  When set to false:
  - Global TCP/UDP intercepts (e.g., telepresence intercept myservice --port 8080) are rejected
  - HTTP intercepts with filters (e.g., --http-header, --http-path-prefix) work normally
  - Multiple developers can create personal HTTP intercepts on the same service simultaneously
  - Users receive a helpful error message directing them to use HTTP-specific flags

  Default behavior:
  - Defaults to true to maintain full backward compatibility
  - All existing intercepts continue to work unchanged

  Implementation Details

  1. Helm Chart (charts/telepresence-oss/):
    - Added intercept.allowGlobalIntercepts to values.yaml with default true
    - Added schema definition to values.schema.yaml
    - Added INTERCEPT_ALLOW_GLOBAL environment variable to templates/deployment.yaml
  2. Traffic Manager (cmd/traffic/cmd/manager/):
    - Added AllowGlobalIntercepts bool field to managerutil.Env struct
    - Added validation in state/intercept.go to check and reject global intercepts when disabled
    - Validation happens early before port lookup for better error messages
  3. Documentation:
    - Added new "Restricting Global Intercepts" section to docs/reference/cluster-config.md

Testing

  Unit Tests:
  - Environment variable parsing (true/false values)
  - Validation logic for enabled/disabled modes
  - HTTP/wiretap intercepts always allowed
  - Error message validation
  - Default behavior verification
  - Replaces should be blocked when allowGlobalIntercepts: false

  Usage Example

  Deploy with global intercepts disabled:

```
telepresence helm install --set intercept.allowGlobalIntercepts=false
```

Attempting a global intercept (will fail):

```
$ telepresence intercept myservice --port 8080
Error: global TCP/UDP intercepts are disabled. Use --http-header or --http-path-* flags for HTTP intercepts
```

Using HTTP intercepts (will succeed):

```
# Developer 1
telepresence intercept myservice --http-header X-User-ID=alice --port 8080
```

```
# Developer 2 (simultaneously)
telepresence intercept myservice --http-header X-User-ID=bob --port 8081
```

Backward Compatibility

Fully backward compatible:
- Defaults to true (existing behavior)
- No changes required to existing deployments
- All existing intercepts continue to work
- Feature is opt-in via Helm configuration


  ---
  cc @thallgren

## Checklist

<!--
  Please review the requirements for each checkbox, and check them
  off (change "[ ]" to "[x]") as you verify that they are complete.
-->

 - [x] I made sure to update `./CHANGELOG.yml` (our release notes are generated from this file).
 - [x] I made sure to add any documentation changes required for my change.
 - [x] My change is adequately tested.
 - [x] I updated `CONTRIBUTING.md` with any special dev tricks I had to use to work on this code efficiently.
 - [ ] Once my PR is ready to have integration tests ran, I posted the PR in #telepresence-oss channel on the
       [CNCF Slack](https://slack.cncf.io/) so that the "ok to test" label can be applied.

bgruszka committed:
Add configuration to control global TCP/UDP intercepts

Introduce a new Helm chart configuration option `intercept.allowGlobalIntercepts` to manage the allowance of global intercepts. When set to `false`, only HTTP intercepts with header or path filters are permitted, enhancing collaboration in shared development environments. Update relevant documentation and tests to reflect this change.

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

bgruszka committed:
Add configuration to disable global TCP/UDP intercepts in Helm chart

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

bgruszka committed:
Refactor test struct formatting in intercept_test.go for improved readability

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

bgruszka committed:
Add integration tests for global intercepts behavior when enabled and disabled

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

bgruszka committed:
Refactor test comments for consistency and clarity in intercept tests

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

thallgren reviewed (commented):
Looks good but I'm missing mentions of the `replace` command. A `replace` will be blocked, so functionality wise everything is OK, but the tests, docs, and error messages need some additional work.

bgruszka committed:
Enhance global intercepts validation to include replace logic and update error messages accordingly

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

bgruszka committed:
Bump version to 2.26.0 and update related documentation

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

bgruszka committed:
Remove integration tests that interfere with other test suites

The integration tests for allowGlobalIntercepts feature were incorrectly
using AddSingleServiceSuite which shares a Traffic Manager instance with
other test suites. Calling TelepresenceHelmInstallOK with upgrade=true
in SetupSuite modified the shared Traffic Manager configuration, causing
the headless service test and potentially other tests to fail.

The feature is already thoroughly tested by unit tests in
cmd/traffic/cmd/manager/state/intercept_test.go which cover all scenarios:
- Global intercepts blocked when disabled
- HTTP intercepts allowed when global disabled
- Wiretaps allowed when global disabled
- Replace operations blocked when disabled

Integration tests would require a custom test suite type to properly
isolate the Traffic Manager configuration, which is not worth the
complexity given the comprehensive unit test coverage.

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

bgruszka commented:
> Looks good but I'm missing mentions of the `replace` command. A `replace` will be blocked, so functionality wise everything is OK, but the tests, docs, and error messages need some additional work.

I've added replace too :) I've also fixed golang-ci-lint errors.

thallgren commented on cmd/traffic/cmd/manager/managerutil/envconfig_test.go:
```diff
@@ -34,7 +34,6 @@ func TestEnvconfig(t *testing.T) {
 		"POD_IP":                          "203.0.113.18",
 		"POD_CIDR_STRATEGY":               "auto",
 		"SERVER_PORT":                     "8081",
-		"INTERCEPT_DISABLE_GLOBAL":        "false",
```
Haha, right. Remnants of the old implementation. Only used by the pro-version.

thallgren commented on CHANGELOG.yml:
```diff
@@ -24,9 +24,21 @@
+          defaults to `true` to maintain full backward compatibility with existing deployments. When a user
+          attempts to create a global intercept while the setting is disabled, they receive a helpful error
+          message suggesting the use of `--http-header` or `--http-path-*` flags for HTTP-filtered intercepts.
+        docs: reference/config#intercept
```
```suggestion
        docs: reference/cluster_config#restricting_global_intercepts
```

bgruszka replied:
I've changed the link :)

thallgren reviewed (approved):
Approving with a suggestion for a better link in the change-logs.

bgruszka committed:
Update documentation link for global intercepts configuration

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>

bgruszka committed:
Update release notes to correct documentation links for global TCP/UDP intercepts configuration

Signed-off-by: Błażej Gruszka <blazej.gruszka@displate.com>
